crates/uv-resolver/src/resolver/mod.rs

```diff
@@ -2057,11 +2046,29 @@ impl ForkState {
                     has_url = true;
                 };
 
-                // `PubGrubDependency` also gives us a local version if specified by the user.
-                // Keep track of which local version we will be using in this fork for transitive
-                // dependencies.
-                if let Some(local) = local {
-                    self.fork_locals.insert(name.clone(), local.clone());
+                // If the specifier is an exact version and the user requested a local version for this
+                // fork that's more precise than the specifier, use the local version instead.
+                if let Some(specifier) = specifier {
+                    // It's possible that there are multiple matching local versions requested with
+                    // different marker expressions. All of these are potentially compatible until we
+                    // narrow to a specific fork.
+                    for local in locals.get(name, &self.markers) {
+                        let local = specifier
+                            .iter()
+                            .map(|specifier| {
+                                Locals::map(local, specifier)
+                                    .map_err(ResolveError::InvalidVersion)
+                                    .and_then(|specifier| {
+                                        Ok(PubGrubSpecifier::from_pep440_specifier(&specifier)?)
+                                    })
+                            })
+                            .fold_ok(Range::full(), |range, specifier| {
+                                range.intersection(&specifier.into())
+                            })?;
+
+                        // Add the local version.
+                        *version = version.union(&local);
```

Comment:
I guess this will just... fail? Since it has to be the union of multiple different `==` requirements, right?

Reply:
This should work fine, we want to allow any compatible local versions before narrowing to a fork.